enzy_htp/structure/structure_region/residue_caps.py

```diff
@@ -0,0 +1,579 @@
+Date: 2024-01-31
+"""
+from __future__ import annotations
+from abc import abstractmethod, ABC
+from copy import deepcopy
+import numpy as np
+from scipy.spatial.transform import Rotation as R
+from typing import List, Tuple, Dict
+
+from ..structure import Structure, Residue, Atom, Chain
+
+from enzy_htp.core.logger import _LOGGER
+from enzy_htp.core.doubly_linked_tree import DoubleLinkedNode
+from enzy_htp.core.math_helper import rotation_matrix_from_vectors
+from enzy_htp import chemical as chem
+
+
+class ResidueCap(Residue, ABC):
+    """The residue cap that link to the corresponding
+    Residue while not having the children atoms actually in
+    that Residue. It ensures that given the defined terminal atom, it can 
+    generate a realistic geometry.
+
+    This is mainly used in handling cases like capping atoms that
+    do not exist in the real structure.
+
+    Attributes:
+        link_residue: The Residue() being capped.
+        link_atom: The terminal Atom() that needs capping specifically.
+        socket_atom: Atom() that connects to the link_atom in the normal Structure().
+        _terminal_type: What type of terminal is it? n-term, c-term, etc.
+        atoms: A List[Atom] of the Atom()'s in the ResidueCap.
+    
+    Properties:
+        idx: The idx of the Residue() being capped.
+        chain: The chain name of the Residue() being capped.
+        name: The residue name of the Residue() being capped.
+        term_type: The property getter of _terminal_type
+        net_charge: The net charge of the ResidueCap() chemical group.
+        """
+    def __init__(self, link_residue: Residue,
+                 link_atom: Atom, socket_atom: Atom,
+                    terminal_type:str):
+        """Simplistic ctor. Will throw if link_atom is NOT in link_residue."""
+        self.link_residue = link_residue
+        self.link_atom = link_atom
+        self.socket_atom = socket_atom
+        if link_atom not in link_residue.atoms:
+            _LOGGER.error("link atom must be one of the link residue atoms")
+            raise ValueError
+
+        self._rtype = chem.ResidueType.RESIDUECAP
+        self._terminal_type = terminal_type 
+
+        if self.term_type == 'nterm':
+            self.atoms = self.get_nterm_atoms()
+        elif self.term_type == 'cterm':
+            self.atoms = self.get_cterm_atoms()
+
+        self.set_ghost_parent()
+
+        self.anchor()
+
+
+    def apply_to_geom(self, geom:Structure) -> ResidueCap:
+        """Create a copy of the current ResidueCap specialization and attach it to the same 
+        link/socket Atom()'s and link Residue() in a different Structure() with all of those
+        structural features. Typically used with a StructureEnsemble.
+
+        Args:
+            geom: The Structure() to apply the ResidueCap() to.
+
+        Returns:
+            A copy of the ResidueCap() specialization applied to the supplied Structure().
+        """
+        result = deepcopy(self) 
```

Comment:
I guess this can be type(self).__new__ or just use the entire constructor by type(self)(sth...)

Reply:
deepcopy in this case is meaningless